From acb686a515e1168bd646060a9121a3ca9c94e0f6 Mon Sep 17 00:00:00 2001
From: Matheus Sampaio Queiroga <srherobrine20@gmail.com>
Date: Thu, 6 Aug 2026 22:50:17 -0300
Subject: [PATCH] dt-bindings: pinctrl: add EcoNet EN751221 pin controller

Document the GPIO, interrupt and pin multiplexing controller
shared by the EN7512, EN7513, EN7521 and EN7526 SoC families.

Signed-off-by: Matheus Sampaio Queiroga <srherobrine20@gmail.com>
---
 .../pinctrl/econet,en751221-pinctrl.yaml      | 225 ++++++++++++++++++
 MAINTAINERS                                   |   1 +
 2 files changed, 226 insertions(+)
 create mode 100644 Documentation/devicetree/bindings/pinctrl/econet,en751221-pinctrl.yaml

diff --git a/Documentation/devicetree/bindings/pinctrl/econet,en751221-pinctrl.yaml b/Documentation/devicetree/bindings/pinctrl/econet,en751221-pinctrl.yaml
new file mode 100644
index 000000000000..0452b931e604
--- /dev/null
+++ b/Documentation/devicetree/bindings/pinctrl/econet,en751221-pinctrl.yaml
@@ -0,0 +1,225 @@
+# SPDX-License-Identifier: (GPL-2.0 OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/pinctrl/econet,en751221-pinctrl.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: EcoNet EN7512/EN7521 Pin Controller
+
+maintainers:
+  - Matheus Sampaio Queiroga <srherobrine20@gmail.com>
+
+description:
+  The pin controller and GPIO block used by the EcoNet EN7512, EN7513,
+  EN7521 and EN7526 SoC families. The first 16 GPIO lines can generate
+  interrupts. GPIOs 2 through 31 also provide a set of SoC-specific muxed
+  functions.
+
+properties:
+  compatible:
+    const: econet,en751221-pinctrl
+
+  airoha,chip-scu:
+    $ref: /schemas/types.yaml#/definitions/phandle
+    description:
+      Phandle to the chip system controller containing the I/O mux register.
+
+  interrupts:
+    maxItems: 1
+
+  gpio-controller: true
+
+  '#gpio-cells':
+    const: 2
+
+  gpio-ranges:
+    maxItems: 1
+
+  interrupt-controller: true
+
+  '#interrupt-cells':
+    const: 2
+
+allOf:
+  - $ref: pinctrl.yaml#
+
+required:
+  - compatible
+  - airoha,chip-scu
+  - interrupts
+  - gpio-controller
+  - '#gpio-cells'
+  - gpio-ranges
+  - interrupt-controller
+  - '#interrupt-cells'
+
+patternProperties:
+  '-pins$':
+    type: object
+
+    patternProperties:
+      '^mux(-|$)':
+        type: object
+        $ref: /schemas/pinctrl/pinmux-node.yaml
+
+        properties:
+          function:
+            enum: [sipo, tod_1pps, pcie_reset, spi, uart, pon, pcm,
+                   pcm_spi, phy_led, gpio]
+
+          groups: true
+
+        required:
+          - function
+          - groups
+
+        allOf:
+          - if:
+              properties:
+                function:
+                  const: sipo
+            then:
+              properties:
+                groups:
+                  enum: [sipo, sipo_rclk]
+          - if:
+              properties:
+                function:
+                  const: tod_1pps
+            then:
+              properties:
+                groups:
+                  enum: [dmt_tod_1pps, pon_tod_1pps, gsw_tod_1pps]
+          - if:
+              properties:
+                function:
+                  const: pcie_reset
+            then:
+              properties:
+                groups:
+                  enum: [pcie_reset0, pcie_reset1]
+          - if:
+              properties:
+                function:
+                  const: spi
+            then:
+              properties:
+                groups:
+                  enum: [spi_quad]
+          - if:
+              properties:
+                function:
+                  const: uart
+            then:
+              properties:
+                groups:
+                  enum: [uart2]
+          - if:
+              properties:
+                function:
+                  const: pon
+            then:
+              properties:
+                groups:
+                  enum: [pon]
+          - if:
+              properties:
+                function:
+                  const: pcm
+            then:
+              properties:
+                groups:
+                  enum: [pcm1, pcm2]
+          - if:
+              properties:
+                function:
+                  const: pcm_spi
+            then:
+              properties:
+                groups:
+                  items:
+                    enum: [pcm_spi, pcm_spi_int, pcm_spi_rst, pcm_spi_cs3,
+                           pcm_spi_cs4]
+                  maxItems: 5
+          - if:
+              properties:
+                function:
+                  const: phy_led
+            then:
+              properties:
+                groups:
+                  enum: [phy0_led, phy1_led, phy2_led, phy3_led, phy4_led]
+          - if:
+              properties:
+                function:
+                  const: gpio
+            then:
+              properties:
+                groups:
+                  items:
+                    enum: [gpio2, gpio3, gpio4, gpio5, gpio6, gpio7,
+                           gpio8, gpio9, gpio10, gpio11, gpio12, gpio13,
+                           gpio14, gpio15, gpio16, gpio17, gpio18, gpio19,
+                           gpio20, gpio22, gpio30, gpio31]
+                  maxItems: 22
+
+        additionalProperties: false
+
+      '^conf(-|$)':
+        type: object
+        $ref: /schemas/pinctrl/pincfg-node.yaml
+
+        properties:
+          pins:
+            items:
+              enum: [gpio0, gpio1, gpio2, gpio3, gpio4, gpio5, gpio6,
+                     gpio7, gpio8, gpio9, gpio10, gpio11, gpio12, gpio13,
+                     gpio14, gpio15, gpio16, gpio17, gpio18, gpio19,
+                     gpio20, gpio21, gpio22, gpio23, gpio24, gpio25,
+                     gpio26, gpio27, gpio28, gpio29, gpio30, gpio31,
+                     gpio32, gpio33, gpio34, gpio35, gpio36, gpio37,
+                     gpio38, gpio39, gpio40, gpio41, gpio42, gpio43,
+                     gpio44, gpio45, gpio46, gpio47, gpio48, gpio49,
+                     gpio50, gpio51, gpio52, gpio53, gpio54, gpio55,
+                     gpio56, gpio57, gpio58, gpio59, gpio60, gpio61,
+                     gpio62, gpio63]
+            minItems: 1
+            maxItems: 64
+
+          input-enable: true
+          output-enable: true
+          output-low: true
+          output-high: true
+
+        required:
+          - pins
+
+        additionalProperties: false
+
+    additionalProperties: false
+
+additionalProperties: false
+
+examples:
+  - |
+    pinctrl: pinctrl {
+      compatible = "econet,en751221-pinctrl";
+      airoha,chip-scu = <&chip_scu>;
+
+      interrupt-parent = <&intc>;
+      interrupts = <10>;
+
+      gpio-controller;
+      #gpio-cells = <2>;
+      gpio-ranges = <&pinctrl 0 0 64>;
+
+      interrupt-controller;
+      #interrupt-cells = <2>;
+
+      uart2-pins {
+        mux {
+          function = "uart";
+          groups = "uart2";
+        };
+      };
+    };
diff --git a/MAINTAINERS b/MAINTAINERS
index b5d67b6c1d10..b6bb34c0784d 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -20334,6 +20334,7 @@ L:	linux-mediatek@lists.infradead.org (moderated for non-subscribers)
 S:	Maintained
 F:	Documentation/devicetree/bindings/mfd/econet,en751221-gpio-sysctl.yaml
 F:	Documentation/devicetree/bindings/pinctrl/airoha,en7581-pinctrl.yaml
+F:	Documentation/devicetree/bindings/pinctrl/econet,en751221-pinctrl.yaml
 F:	drivers/pinctrl/airoha/pinctrl-airoha.c
 
 PIN CONTROLLER - AMD
-- 
2.55.0

